### PR TITLE
Obtain base_url in a more robust and standard way

### LINKED ISSUE #146: Fix base URLs.
When tiled is served behind gunicorn and/or nginx, it provides the wrong base URL in the `links` it hands the client. Example:

```
$ http https://tiled-demo.blueskyproject.io/metadata/bmm | jq .data.links
{
  "self": "http://localhost:8000/metadata/bmm",
  "search": "http://localhost:8000/search/bmm"
}
```

Somehow we need to propagate the relevant information through. Off the top of my head, maybe this could involve...

* Grab it from the `Host` header (seems simplest if it is a general solution)
* Do something with `X-Forwarded-For`
* Extract something from the ASGI protocol

### PR DESCRIPTION
**Pull request by danielballan**
Closes #146 

To test, I deployed this on https://tiled-demo.blueskyproject.io/, which serves Tiled using gunicorn behind nginx. The scheme and the host are both correct:

```
$ http https://tiled-demo.blueskyproject.io/node/metadata/ | jq .data.links
{
  "self": "https://tiled-demo.blueskyproject.io/node/metadata/",
  "search": "https://tiled-demo.blueskyproject.io/node/search/"
}
```

We also care about the niche use case of SSH port-forwarding to access a Tiled deployment that is only available on an internal network (such as `tiled.nsls2.bnl.gov` is now). HTTPS over port-forwarding is janky (it breaks SSL) but it has been useful in the early stages. Fortunately, it is supported by the approach in this PR:

```
$ ssh -L 8000:localhost:443 tiled-demo.blueskyproject.io
...
$ http --verify false "https://localhost:8000/node/metadata/" | jq .data.links
{
  "self": "https://localhost:8000/node/metadata/",
  "search": "https://localhost:8000/node/search/"
}
```

It works because the `Host` header in the request is set to `localhost:8000` and propagated through nginx as `X-Forwarded-Host`.

I had previously supported this use case with a custom `x-base-url` header, set by the Python client and observed by the server. This is not necessary, and it is removed in this PR.

### CODE REVIEW

File: tiled/server/router.py [hunk 1]
```diff
@@ -465,19 +465,26 @@ def node_full(
...
+    #   https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.23
+    host = request.headers.get("x-forwarded-host", request.headers["host"])
+    scheme = request.headers.get("x-forwarded-scheme", request.url.scheme)
     root_path = request.scope.get("root_path") or "/"
-    if isinstance(url.netloc, bytes):
-        netloc_str = url.netloc.decode()
-    else:
-        netloc_str = url.netloc
-    return f"{url.scheme}://{netloc_str}{root_path}"
+    return f"{scheme}://{host}{root_path}"
```
Comment (dylanmcreynolds) [inline tiled/server/router.py:490]: This line depends on root_path, if present, always starting with a `/`. Is that always dependable?

File: tiled/server/router.py [hunk 2]
```diff
@@ -465,19 +465,26 @@ def node_full(
...
+    #   set by default. With nginx, for example, they need to be configured.
+    #
+    # * The client may be connecting through SSH port-forwarding. (This
+    #   is a niche use case but one that we nonetheless care about.)
+    #   The Host or X-Forwarded-Host header may include a non-default port.
+    #   The HTTP spec specifies that the Host header may include a port
+    #   to specify a non-default port.
+    #   https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.23
+    host = request.headers.get("x-forwarded-host", request.headers["host"])
+    scheme = request.headers.get("x-forwarded-scheme", request.url.scheme)
```
Comment (dylanmcreynolds) [inline tiled/server/router.py:488]: This looks wrong. The header is `x-forwarded-proto`, not `x-forwarded-scheme`

**Review by dylanmcreynolds — CHANGES_REQUESTED** [review]

File: tiled/server/router.py (hunk 2 above)
Comment (dylanmcreynolds) [reply tiled/server/router.py:488]: Also, can we depend on headers at this point being lower case? HTTP headers are not case sensitive.

**Review by dylanmcreynolds — CHANGES_REQUESTED** [review]
I have some thoughts to look at.

**Review by dylanmcreynolds — CHANGES_REQUESTED** [review]
Added a couple comments I'd like to review.

File: tiled/server/router.py (hunk 2 above)
Comment (danielballan) [reply tiled/server/router.py]: Oops, I misspelled this consistently in the nginx config and here, which is why the test worked. Fixed.

The `headers` object is a special case-insensitive dict.

File: tiled/server/router.py (hunk 1 above)
Comment (danielballan) [reply tiled/server/router.py:492]: It looks like there are no guarantees that it has a trailing slash, so I have added a check that ensures it gets one.

**Review by dylanmcreynolds — APPROVED** [review]

**Pull request merged**